lucene/core/src/java/org/apache/lucene/codecs/KnnVectorsWriter.java

```
@@ -303,29 +314,45 @@ private MergedFloat32VectorValues(List<FloatVectorValuesSub> subs, MergeState me
       }
 
       @Override
-      public int docID() {
-        return docId;
-      }
+      public DocIndexIterator iterator() {
+        return new DocIndexIterator() {
+          private int index = -1;
 
-      @Override
-      public int nextDoc() throws IOException {
-        current = docIdMerger.next();
-        if (current == null) {
-          docId = NO_MORE_DOCS;
-        } else {
-          docId = current.mappedDocID;
-        }
-        return docId;
-      }
+          @Override
+          public int docID() {
+            return docId;
+          }
 
-      @Override
-      public float[] vectorValue() throws IOException {
-        return current.values.vectorValue();
+          @Override
+          public int index() {
+            return index;
+          }
+
+          @Override
+          public int nextDoc() throws IOException {
+            current = docIdMerger.next();
+            if (current == null) {
+              docId = NO_MORE_DOCS;
+              index = NO_MORE_DOCS;
+            } else {
+              docId = current.mappedDocID;
+              ++index;
+            }
+            return docId;
+          }
+
+          @Override
+          public int advance(int target) throws IOException {
+            throw new UnsupportedOperationException();
+          }
+        };
       }
 
       @Override
-      public int advance(int target) {
-        throw new UnsupportedOperationException();
+      public float[] vectorValue(int ord) throws IOException {
+        // FIXME what can we assert here?
+        // assert ord == iterator.index();
+        return current.values.vectorValue(current.index());
```

This is the biggest "gotcha" in this whole thing I think. Does `DocIdMerger` allow random access at all?

It seems like `vectorValue(ord)` should also be able to jump between `current` sub iterators and move backwards and forwards. But, I don't think `DocIdMerger` allows backwards movement at all.

---

I think unless we can fix `DocIdMerger`, we should throw an error here indicating that only forward iteration is allowed.

---

I had thought we'd fix it by concatenating the vector dictionaries of each segment. Then based on the requested vector ordinal, you could compute the segment that the vector belongs to via `ReaderUtil#subIndex`.

---

There is something similar here in `SlowCompositeCodecReaderWrapper` - binary search across sub-values